### PR TITLE
Fix Oracle PList generation

### PR DESCRIPTION
**Pull request by tofi86**
This PR contains changes made after the discussion in #8.

* 282fab8 Fix parsing of `jvmversion` to support Java 9+ (new *semver* versioning scheme)
* 0f68620 Introduce a new attribute `useOracleStyle` as an alternative to `useJavaXKey`

**What does the PR fix?**

With PR #8 @Vampire introduced and @tobium merged code changes that prevented the use of the `JavaX` key.

The new code would always use the Oracle PList syntax when a `jvmversion` > 1.7 was requested. However, with the `JavaX` key feature and e.g. the `universalJavaApplicationStub` as a launcher stub you can perfectly create App bundles wich run with Java 7, 8+. No need to fall back to the Oracle Plist style.

**What is the `JavaX` key for anyways?**

The JavaX key isn't standardized. It's supported by https://github.com/tofi86/universalJavaApplicationStub (~100 stars, ~30 clones a day), a project I created when Oracle took over Java development from Apple. The purpose was to have a launcher stub supporting the "old" `Java` key even on Java 7 and 8 because the new Oracle PList style didn't support all features.
Then with Java 8 Oracle started to show warning messages and redirecting Apps using the `Java` key to the old Java 6. Together with the original developer of JarBundler we came up with the idea of using a non-standard key named `JavaX` and I did the implementation in both projects (first in my fork of JarBundler, then merged here).

This is still relevant because the Oracle PList style didn't evolve (in contrary: Oracle discontinued their appbundler project) and the syntax still doesn't support all the features the Apple dictionary did. Call it "legacy" mode. In the meantime quite a bunch of tools use this feature frequently: https://github.com/search?q=usejavaxkey&type=Code

That's why the `useJavaXKey` attribute exists and should continue working in this project.

**Why is it implemented like it is?**

The changes from #8 were were merged after the 3.3 release but were never officially released on Maven Central.

Therefore I chose to refactor the Oracle Plist style implementation by @Vampire in a similar way to the `JavaX` implementation: As an additional attribute: `useOracleStyle`.

Here are a couple of annotated examples:

```xml
<jarbundler jvmversion="1.7+" …>
```
* Before (with PR #8): *Switched to the Oracle Plist style without notification in the build log*
* Now: *Continues to work as in v3.3.0. No auto switch to the Oracle Plist style. Creates the Apple Plist style with old `Java` dictionary.*

----

```xml
<jarbundler jvmversion="1.7+" useJavaXKey="true" …>
```
* Before (with PR #8): *Failed with Exception `Setting usejavaxkey is useless if jvmversion is at least 1.7, because then the Oracle PList format is used`*
* Now: *Continues to work as in v3.3.0. No auto switch to the Oracle Plist style. Creates the Apple Plist style in "legacy mode" with `JavaX` dictionary.*

----

```xml
<jarbundler jvmversion="1.7+" useJavaXKey="true" useOracleStyle="true" …>
```
* Before (with PR #8): *Failed with Exception `Setting usejavaxkey is useless if jvmversion is at least 1.7, because then the Oracle PList format is used`*
* Now: *Failes with Exception `Setting attribute 'useOracleStyle' is useless if attribute 'useJavaXKey' is used`*

----

```xml
<jarbundler jvmversion="1.7+" useOracleStyle="true" useJavaXKey="true" …>
```
* Before (with PR #8): *Failed with Exception `Setting usejavaxkey is useless if jvmversion is at least 1.7, because then the Oracle PList format is used`. *
* Now: *Failes with Exception `Setting attribute 'useJavaXKey' is useless if attribute 'useOracleStyle' is used`*

----

```xml
<jarbundler jvmversion="1.7+" useOracleStyle="true" …>
```
* Now: *Creates the Oracle Plist style.*

----

**Conclusion:**

With these changes we maintain backward compatibility to v3.3 and do not break existing build scripts due to explicit auto switch to Oracle Plist style.

However, if you'd like to use the Oracle PList format you can now set this option explicitly.

----

@Vampire @tobium I'd like to hear your comments and thoughts on that :-)

I'm very much looking forward to see this released as v3.4.0!!

### CODE REVIEW

**Comment by tofi86** [comment]
Hi @tobium, any news here? What are your thoughts about this?

**Comment by tofi86** [comment]
Hey guys, this has been open for over a month now. I would like to merge this soon if there are noe objections.

**Review by Vampire — APPROVED** [review]
Sorry for the long time until I answer and sorry for destroying the JavaX functionality.
I guess I was not aware of your shell stub or that the JavaX key is for that one, or read somewhere that it is from the Apple Java.
Now that I had to look at this again as the Oracle stub was removed between Java 8 and 9 (seems those guys didn't know what it is for and just removed it from the sources (https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8048337 / https://github.com/openjdk/jdk/commit/52db401053b8d3556a9c7f2f276be5bf00c82e55) and the old one is not compatible with new Java versions anymore, I understand better.

I second your changes to decide manually which format to use and whether to use `Java` or `JavaX`.
Maybe it would be better though to use some "enumerated" value with three options (like `Apple Stub`, `Oracle Stub` and `Universal Stub`) instead of two separate booleans that exclude each other?

Btw. regarding "Where is the classpath entry? where is the JVMVersion key? where the splashscreen key?" and "The Oracle Plist format generated by this PR needs to be fixed", there is nothing to fix, they are missing because they are not supported. Or at least when I made this PR only the three keys output were supported. Now it is differently, the original project is more or less dead and the fork that is still maintained added a bunch of additional possibilities.
Back then the according Appbundler Ant task was very limited and missed some of the possibilities of JarBundler, but it might even have swapped now, as you can even add arbitrary plist entries with the current Appbundler task, so it is probably preferable to actually use Appbundler now instead of JarBundler, especially if you use their application stub. Up to now I used their launcher stub and JarBundler to assemble the package.

**Comment by tofi86** [comment]
> Sorry for the long time until I answer and sorry for destroying the JavaX functionality.
> I guess I was not aware of your shell stub or that the JavaX key is for that one, or read somewhere that it is from the Apple Java.

nevermind...

> Now that I had to look at this again as the Oracle stub was removed between Java 8 and 9 (seems those guys didn't know what it is for and just removed it from the sources (https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8048337 / [openjdk/jdk@52db401](https://github.com/openjdk/jdk/commit/52db401053b8d3556a9c7f2f276be5bf00c82e55)) and the old one is not compatible with new Java versions anymore, I understand better.

d'oh! I wasn't aware that they completely removed it from sources, though I was aware that the AppBundler project which distributed it was discontinued.

> I second your changes to decide manually which format to use and whether to use `Java` or `JavaX`.
> Maybe it would be better though to use some "enumerated" value with three options (like `Apple Stub`, `Oracle Stub` and `Universal Stub`) instead of two separate booleans that exclude each other?

I'm fine with that as well and could change the PR.
_However_, we would break backward compatibility with removing the `useJavaXKey` key and replacing it with a completely new solution.
What do you think?

> Btw. regarding "Where is the classpath entry? where is the JVMVersion key? where the splashscreen key?" and "The Oracle Plist format generated by this PR needs to be fixed", there is nothing to fix, they are missing because they are not supported. Or at least when I made this PR only the three keys output were supported. Now it is differently, the original project is more or less dead and the fork that is still maintained added a bunch of additional possibilities.

Yep, got it. Just wanted to note that with "removing" the old `Java` key syntax and switching completely to the Oracle style syntax, some options would get lost.

**Comment by Vampire** [comment]
> > Now that I had to look at this again as the Oracle stub was removed between Java 8 and 9 (seems those guys didn't know what it is for and just removed it from the sources (https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8048337 / [openjdk/jdk@52db401](https://github.com/openjdk/jdk/commit/52db401053b8d3556a9c7f2f276be5bf00c82e55)) and the old one is not compatible with new Java versions anymore, I understand better.
> 
> d'oh! I wasn't aware that they completely removed it from sources, though I was aware that the AppBundler project which distributed it was discontinued.

Nah, that's just me telling you bullshit, sorry.
After I wrote this I remembered properly and forgot I wrote this non-sense here.
What was removed between Java 8 and Java 9 were the sources for the old Apple stub that were still in the OpenJDK sources.
The Oracle replacement was the one at http://java.net/projects/appbundler/.
That original source is dead of course as java.net was discontinued.
So the fork of TheInfiniteKind at https://github.com/TheInfiniteKind/appbundler
and its depending project that builds it to releases and also releases it to Maven Central are the actively maintained successors.

> I'm fine with that as well and could change the PR.
> _However_, we would break backward compatibility with removing the `useJavaXKey` key and replacing it with a completely new solution.
> What do you think?

Actually now that the AppBundler has developed, so that you can also set the stuff you couldn't when I made my PR here like the additional plist entries for high resolution and so on, I would also be ok with reverting my PR, as it does not fully support all their capabilities and as I changed our build to also use the AppBundler task, not only their stub, I'm not going to make any further PRs to support their format here.

**Comment by tofi86** [comment]
> Actually now that the AppBundler has developed, so that you can also set the stuff you couldn't when I made my PR here like the additional plist entries for high resolution and so on, I would also be ok with reverting my PR, as it does not fully support all their capabilities and as I changed our build to also use the AppBundler task, not only their stub, I'm not going to make any further PRs to support their format here.

If it's okay with you, we should revert it. Will you push the revert? Shall I revert it and push to master?

Is there anyone else who maintains the repository and would like to review? Do you know about @tobium?

**Comment by slajar** [comment]
@tofi86 I can also review pullrequests on this repo.

**Comment by Vampire** [comment]
I cannot push, I can only open PRs.
Feel free to push or make a PR that completely removes my PR changes.

**Comment by tofi86** [comment]
It's been a while... 😅 But finally, I managed to revert the changes and send the PullRequest.

Once this is merged, I'm going to close this PR and rebase some of its commits on the new `master`:

* 282fab8540f6467d509b99e0fdad93c140a1f2e1 support Java 9, 10, etc. in 'jvmversion' attribute
  * _also requested in #16 lately, which caught my attention again..._
* 21f03ae7c915ce3842d8a507e02a8f5e4887a554 update .gitignore
* 3991d6b4f9f61f1827134e90fb1775883d961dfc fix typos in original code

I will drop the commit I added with this PR which introduced the `useOracleStyle` key (0f68620ca9c0b3a6fdcf60c96576d783d29ed6c9).

**Pull request closed without merging**
